host/sgx/linux/vdso.c

```diff
@@ -0,0 +1,205 @@
+
+    switch (run->function)
+    {
+        case ENCLU_EENTER:
+            /* Unexpected case (e.g., the enclave loses EPC context
+             * because of power events). Return failing value. */
+            result = -1;
+            break;
+        case ENCLU_EEXIT:
+        {
+            /* Regular exit (the enclave finishes an ECALL or makes an
+             * OCALL). Return zero.
+             * Note that an alternative implementation is returning
+             * ENCLU_EENTER. However, doing so requires setting up
+             * the input parameters into corresponding registers (e.g.,
+             * rdi, rsi, and rdx) and ensuring the compiler to preserve
+             * these registers until the function returns. Instead,
+             * we return zero to avoid dealing with such complexities
+             * and also to use similar implementation as regular enter. */
+            return_args->rdi = arg1;
+            return_args->rsi = arg2;
+            result = 0;
+            break;
+        }
+        case ENCLU_ERESUME:
+        {
+            /* Hardware exceptions occur */
+
+            oe_host_exception_context_t host_context = {0};
+
+            host_context.rax = ENCLU_ERESUME;
+            host_context.rbx = run->tcs;
+
+            /* AEP is assigned by vDSO implementation */
+
+            oe_host_handle_exception(&host_context);
+
+            result = ENCLU_ERESUME;
+            break;
+        }
+    }
+
+done:
+    /* If the result <= 0, the value will be forwared as the return
+     * value of _vdso_sgx_enter_enclave. Otherwise, _vdso_sgx_enter_enclave
+     * will invoke the ENCLU[result] instead of returning to the caller. */
+    return result;
+}
+
+oe_result_t oe_vdso_enter(
+    void* tcs,
+    uint64_t arg1,
+    uint64_t arg2,
+    uint64_t* arg3,
+    uint64_t* arg4,
+    oe_enclave_t* enclave,
+    oe_ecall_context_t* ecall_context)
+{
+    oe_result_t result = OE_UNEXPECTED;
+    struct sgx_enclave_run run = {0};
+    vdso_args_t return_args = {0};
+    int return_value = 0;
+    uint32_t mxcsr = 0;
+    uint16_t fcw = 0;
+
+    run.tcs = (uint64_t)tcs;
+    run.user_handler = (uint64_t)oe_vdso_user_handler;
+    run.user_data = (uint64_t)&return_args;
+
+    while (1)
+    {
+        // Compiler will usually handle this on exiting a function that uses
+        // AVX, but we need to avoid the AVX-SSE transition penalty here
+        // manually as part of the transition to enclave. See
+        // https://software.intel.com/content/www/us/en/develop/articles
+        // /avoiding-avx-sse-transition-penalties.html
+        if (oe_is_avx_enabled)
+            OE_VZEROUPPER;
+
+        asm volatile("stmxcsr %[mxcsr] \n\t" // Save MXCSR
+                     "fstcw %[fcw] \n\t"     // Save x87 control word
+                     :
+                     : [fcw] "m"(fcw), [mxcsr] "m"(mxcsr)
+                     :);
+
+        return_value = (_vdso_sgx_enter_enclave)(
```

Comment:
`_vdso_sgx_enter_enclave` is not x64 ABI conformant. How/Where do we ensure that the register are correctly preserved at this call? Need comments if we are already doing it

Reply:
According to the comments in sgx.h:

>   NOTE: __vdso_sgx_enter_enclave() does not ensure full compliance with the
  x86-64 ABI, e.g. doesn't handle XSAVE state.  Except for non-volatile
  general purpose registers, EFLAGS.DF, and RSP alignment, preserving/setting
  state in accordance with the x86-64 ABI is the responsibility of the enclave
  and its runtime, i.e. __vdso_sgx_enter_enclave() cannot be called from C
  code without careful consideration by both the enclave and its runtime.

We don't need to handle general propose registers (i.e., RBX, RSP, RBP, R12-15).
Other than those, we did the same as regular oe_enter implementation:
- clear YMM, save/restore MXCSR and FPU control word